data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanOptions.java

```diff
@@ -99,36 +83,62 @@ public Builder setRange(Duration range) {
             return this;
         }
 
-        public Builder setBucket(String bucket) {
-            this.bucket = bucket;
+        public Builder setEndDateTime(LocalDateTime endDateTime) {
+            this.endDateTime = endDateTime;
             return this;
         }
 
-        public Builder setEndDateTime(LocalDateTime endDateTime) {
-            this.endDateTime = endDateTime;
+        public Builder setBucketOption(S3ScanBucketOption bucketOption) {
+            this.bucketOption = bucketOption;
             return this;
         }
-        public ScanOptions build() {
-            if(Objects.nonNull(startDateTime) && Objects.nonNull(range) && Objects.nonNull(endDateTime))
-                scanRangeDateValidationError();
 
-            if(Objects.nonNull(startDateTime) && Objects.nonNull(endDateTime)){
-                this.useStartDateTime = startDateTime;
-                this.useEndDateTime = endDateTime;
-            } else if (Objects.nonNull(endDateTime) && Objects.nonNull(range)) {
-                this.useStartDateTime = endDateTime.minus(range);
-                this.useEndDateTime = endDateTime;
-            } else if(Objects.nonNull(startDateTime) && Objects.nonNull(range)) {
-                this.useStartDateTime = startDateTime;
-                this.useEndDateTime = startDateTime.plus(range);
-            } else
+        public ScanOptions build() {
+            LocalDateTime bucketStartDateTime = Objects.isNull(bucketOption.getStartTime()) ? startDateTime : bucketOption.getStartTime();
+            LocalDateTime bucketEndDateTime = Objects.isNull(bucketOption.getEndTime()) ? endDateTime : bucketOption.getEndTime();
+            Duration bucketRange = Objects.isNull(bucketOption.getRange()) ? range : bucketOption.getRange();
+
+            long nonNullCount = Stream.of(bucketStartDateTime, bucketEndDateTime, bucketRange)
+                    .filter(Objects::nonNull)
+                    .count();
+
+            if (nonNullCount == 0 || nonNullCount == 2) {
+                setDateTimeToUse(bucketStartDateTime, bucketEndDateTime, bucketRange);
+            } else if (nonNullCount == 3) {
+                long originalBucketLevelNonNullCount = Stream.of(
+                        bucketOption.getStartTime(), bucketOption.getEndTime(), bucketOption.getRange())
+                        .filter(Objects::nonNull)
+                        .count();
+
+                if (originalBucketLevelNonNullCount == 2) {
+                    setDateTimeToUse(bucketOption.getStartTime(), bucketOption.getEndTime(), bucketOption.getRange());
+                } else {
```

Comment:
Everything looks really good overall! I think there is only edge case missing

```
start_time: "start"
range: "my-range"
bucket:
   - name: "my_bucket"
      end_time: "custom_end"
```

I think this would be considered invalid with the current code, so we could have an extra branch here 

```
else if (originalBucketLevelNonNullCount == 1) {
    if (Objects.nonNull(bucketOption.getStartTime() || Objects.nonNull(bucketOption.endTime()) {
           bucketRange = null;
    } else { // means range only is used on the bucket
            endTime = null;
    }

    setDateTimeToUse(startTime, endTime, durationRange);
}
```

Reply:
I consider the cases where two of start_time, end_time, range on the global level and the third one on bucket level invalid because there's ambiguity. For example in this case:
```
start_time: "start"
range: "my-range"
bucket:
   - name: "my_bucket"
      end_time: "custom_end"
```
Either (start_time, end_time) or (range, end_time) can be used to define the time range.

Reply:
Ah yes I thought end_time minus range is equivalent to start_time and end_time, but it is not.